src/ir/auth_logic/ast.h

```diff
@@ -76,34 +77,34 @@ class Attribute {
 class CanActAs {
  public:
   explicit CanActAs(Principal left_principal, Principal right_principal)
-    : left_principal_(std::move(left_principal)),
-      right_principal_(std::move(right_principal)) {}
+      : left_principal_(std::move(left_principal)),
+        right_principal_(std::move(right_principal)) {}
   const Principal& left_principal() const { return left_principal_; }
   const Principal& right_principal() const { return right_principal_; }
 
  private:
-   Principal left_principal_;
-   Principal right_principal_;
+  Principal left_principal_;
+  Principal right_principal_;
 };
 
-// BaseFact corresponds to a base-case fact that cannot recursively include other 
-// facts through "canSay". "Fact" includes the recursive case. It is necessary 
-// that BaseFact is a separate class from Fact because the RHS of conditional 
-// assertions can only contain BaseFacts.
+// BaseFact corresponds to a base-case fact that cannot recursively include
+// other facts through "canSay". "Fact" includes the recursive case. It is
+// necessary that BaseFact is a separate class from Fact because the RHS of
+// conditional assertions can only contain BaseFacts.
 // Base facts have three forms:
 //  - A predicate
 //  - An attribute
 //  - A canActAs expression
-// The particular choice of form is represented with a variant which is 
-// encapsulated.
+using BaseFactVariantType = std::variant<Predicate, Attribute, CanActAs>;
```

Comment:
Putting public aliases in the header in this fashion is not allowed by the style guide, unfortunately. See https://google.github.io/styleguide/cppguide.html#Aliases . You could put this `using` alias in the `private` section of the class definition, but that wouldn't help you with not repeating the type.

Reply:
From what I can tell, the decision in the style guide is that aliases can't be public in a header unless they are meant to be a part of the public API. In this case, I think it's fair to say that these types _are_ a part of the public API because here variants are traversed by `std::visit`s in compiler passes that operate on the AST. This might seem strange because it might seem like the variant should be encapsulated, but really I think there is no need to encapsulate this since the variant just represents the alternative forms for this piece of syntax. I added a comment to document how these aliases are meant to be used, like the style guide suggests should be done in these cases.

Reply:
(Leaving this one open to see if you agree :)

Reply:
I guess I could buy that, especially with the errors that it prevents in restating the type. I still do feel like this alias should be encapsulated though; perhaps you could make it a public member of the class in which it appears?

Reply:
ok done.